### ISSUE TITLE
DIST packaging changes note formatting

### ISSUE THREAD

**Issue by cv-on-hub** [finale-lua/lua-scripts#604]
In the larger scheme of things this is a trifling problem but I thought I'd mention it anyway.

I use `finaleplugin.Notes` as the text source for INFO pages in dialogs. These are carefully formatted for the best in-dialog display but the formatting gets chewed up in the DIST packaging - even just by deleting carefully placed blank lines. Don't know why these  line deletions in the SRC files are desirable - the file size saving is negligible. 

... so if that process is getting updated any time ...

**Comment by rpatters1** [comment]
The `finaleplugin.Notes` field is markdown, or at least a subset of markdown. It doesn't seem like a good fit for displaying raw to users, whether processed or not. If anything, I hope the future moves towards a more complete implementation of markdown for `finaleplugin.Notes`, which will make it even less suitable for displaying raw to a user.

The next version of RGP Lua will be introducing a new `FCCtrlTextEditor` control. It is a full-blown text display and editing control that can take rich text or enigma strings and format them for display or editing. (You can then extract the edited versions, with formatting.) If you want to display an info page, that's what I would recommend you use. Maybe you could design it in an editor and export the rich text to be included in your script or config data. It may even be able to display embedded graphics. I haven't tested it, but it might work.

I will be introducing two scripts that make use of a fairly wide range of the capabilities of the control. One is a Lyrics Hyphenator that basically replaces the Lyrics Window but includes ChatGPT prompts for automatic hyphenation. The other is (finally) a simple programming console for ad hoc scripts or just to run a script that isn't configured. It basically has all the most important features of JW Lua's dev window plus some additional goodies.

**Comment by cv-on-hub** [comment]
`FCCtrlTextEditor` and all the other enhancements for RGP 0.67+ sound great. An ever-bigger leap forward!

I'm using `finaleplugin.Notes` for INFO pages to avoid duplicating the text, and never use Markdown properties, except forcing the odd line break. (Plain text is fine for my purposes). But that remains a current problem within `finaleplugin.Notes` (at least as rendered on `finalelua.com`) because the DIST packaging process is removing empty lines in all multiline strings, including those within `plugindef()`, so wrecking any Markdown employed therein. It's been doing that since April or so, though not (AFAICT) on older files.

**Comment by jwink75** [comment]
Wow… that sounds amazing!!

**Comment by rpatters1** [comment]
I see these issues:

- Scripts in this repo are part of the repo and so should not be using developer-specific conventions. A future contributor who does not know about the INFO window convention could end up introducing markdown into the `Notes` that messes up the INFO window.
- The `Notes` *should* be using markdown. (And indeed, I looked at some of @Carls's and they do at least use ticks for code blocks.) The whole point is to make the site more inviting and informative to users. Restricting yourself to the subset that works for both the site and an INFO window means no text formatting in either.
- Who are the INFO windows for? Is it really necessary to duplicate the info in the script and on the website? If it is important to have it both places, maybe we should build that into all of the scripts using some sort of convention. Or, alternatively, would it be possible to simply launch a link? One of the small annoyances of the current site is that there is no way to deep-link to a specific script's online notes. I think that is a mod we should look at as well, if we ever get the site to build again.
- Personally, I think an INFO window inside the script is going to need different text than the markdown in the `Notes`, and trying to keep them the same will hurt both formats.

If we ever get the site to build again, we can look into not modifying the Notes. I'm pretty sure the code right now is compressing wherever possible, just because the `dist` scripts get really big. But for now I'd just like to see the site build at all. Ultimately a better solution might be to come up with a convention for *all* scripts and use a tool like `pandoc` to convert the notes into a format that is appropriate for user display inside a script.

**Comment by cv-on-hub** [comment]
Good point all @rpatters1. Just to say that judging by FB feedback, some (many?) of our users don't bother to read `plugindef.Notes`, and presenting that *Info* right in the user dialog is massively convenient for everyone, especially in scripts with many moving parts like my [measure_span.lua](https://github.com/finale-lua/lua-scripts/blob/master/src/measure_span.lua). One benefit of `Markdown` is that reading \*\*bold\*\* isn't much worse than reading **bold**, and in-dialog Notes don't need to be pretty just informative. But retaining correct line breaks is \*\*\_really important!\_\*\*

**Comment by cv-on-hub** [comment]
There probably isn't much else to say here so if there's no other input I'll close this issue in a few days.

**Comment by rpatters1** [comment]
I would leave it open. If the website ever gets fixed, we might want to do something about it.

**Comment by cv-on-hub** [comment]
In case this ever gets fixed, I also flag that trailing spaces on every line are being cropped, which also destroys Markdown formatting, as well as my attempts to retain line formats in internal "info_notes".

**Comment by asherber** [comment]
@cv-on-hub Can you point me to a specific example of these changes in src/dist?

**Comment by cv-on-hub** [comment]
Compare [/src/cross_staff_offset.lua](https://github.com/finale-lua/lua-scripts/blob/master/src/cross_staff_offset.lua#L17) (lines 17-21) with [dist/cross_staff_offset.lua](https://github.com/finale-lua/lua-scripts/blob/master/dist/cross_staff_offset.lua#L4966) (lines 4966-4969).  Blank lines and spaces at the end of every line are removed. Which is true in every line of every `dist` file.

The same is true for all multi-line text selections so I now take preventive measure to retain text spacing for "Info" listings in `dist` files. As here: [/dist/cue_notes_overlay.lua](https://github.com/finale-lua/lua-scripts/blob/master/dist/cue_notes_overlay.lua#L5291)

**Comment by asherber** [comment]
Okay, this looks like a (possibly unintentional) side effect of a function in the bundler that removes comments. It's probably possible to make the bundler smarter so that it leaves Notes alone.

In many scripts, the Notes are just a description of what the script does; others use the Notes for more detailed usage instructions. Since the Notes can use Markdown, I think it would be nice to be able to use them to display formatted text in the new `FCCtrlTextEditor` that Robert has introduced.

I could imagine modifying the bundler so that it takes the Notes and does Markdown to HTML to RTF, which I think could then be used in the editor control. There could be a library function to facilitate this -- basically encapsulating what Carl does -- but no script would be required to make use of it.

**Comment by cv-on-hub** [comment]
I understand the bundler is trying to keep file size down, and `dist` files don't really need to be human-readable, but losing a byte here and there on trailing spaces and 2 bytes for each blank line seems to lose a lot more than it gains. If it's specifically aiming at extended comments surely it could aim more precisely at "--" and "[["..."]]"?

**Comment by rpatters1** [comment]
What tool would we use to convert markdown to RTF? Does `pandoc` do it? I'm wondering if there's a tool that would be easy to embed. If `Notes` will be auto-converted from markdown to rtf by the build, it would be difficult to test what it looks like from the src without some integrated tool.

**Comment by asherber** [comment]
I'm not sure, just started poking around. I think the likeliest route is md->html->rtf. There are lua libraries for the first part, but likely not for the second. 

I haven't worked with `FCCtrlTextEditor` -- is it possible to enhance it so that it can also render HTML, or is that an big lift?

**Comment by rpatters1** [comment]
I think it's a big lift. The OS controls have built-in support for RTF, but not (sfaik) html.

**Comment by asherber** [comment]
The Markdown in Notes should be fairly simple: bold, italic, lists, headings, code, maybe tables. In the website repo, I recently made a change so that the Markdown parser spits out exactly the HTML we want, with classes. I could use the same technique to have the parser emit RTF instead of HTML. I wouldn't generally want to have to write RTF by hand, but the RTF needed for this reasonably constrained subset of HTML wouldn't be that big a deal.

The parser for the website is in JS. That could be incorporated into the build process for the scripts, but as Robert suggests, that makes it harder to preview. It looks like there is also a [parser in lua](https://github.com/jgm/lunamark) which similarly allows for modification of the emitted text. I'll play around with this a little, see what I can get working.

**Comment by asherber** [comment]
Hmm, it has several dependencies, probably not a great candidate for embedding in RGPLua. I'll do a POC with the JS parser, just to see if I can get RTF working.

**Comment by rpatters1** [comment]
The `pandoc` utility lists rtf as an output format. I haven't tried it, but if it works, it should  be easy for a developer to get a preview with that by running it with os.execute against the contents of the Notes.

**Comment by asherber** [comment]
Actually, pandoc does a nice, minimal md to rtf. We could just use that in the build process.

**Comment by asherber** [comment]
I have a working POC of this, with notes converted to RTF during the build process.

![image](https://github.com/finale-lua/lua-scripts/assets/5248041/739ab348-2c7a-43e1-8442-33c3166841f5)

@rpatters1 How do I get the text control to have nothing selected and scroll to the top? In my code (largely cribbed from your sample, which has the same issue), I wind up with everything selected and scrolled to the bottom.

```lua
local function show_notes_dialog()
    local dlg = finale.FCCustomLuaWindow()
    dlg:SetTitle(finale.FCString("Notes Test"))
    local edit_text = dlg:CreateTextEditor(10, 10)
    edit_text:SetWidth(400)
    edit_text:SetHeight(180)
    edit_text:SetUseRichText(true)
    edit_text:SetReadOnly(true)
    edit_text:SetWordWrap(true)
    
    dlg:RegisterInitWindow(
        function()
            edit_text:SetRTFString(finale.FCString(finaleplugin.Notes))
            edit_text:SetSelection(finale.FCRange(0, 0))
            edit_text:ScrollToTop()
        end)
    dlg:ExecuteModal(nil)
end
```

**Comment by rpatters1** [comment]
This is a "feature" of windows that when you tab into the field, it selects the entire text, which scrolls it to the bottom. And when there is only one field, that's where it goes first. I'm not sure how much I can do. Maybe offer an options to make the text un-selectable. But that's fairly user-hostile.

Add a control ahead of it, and see if the problem doesn't go away.

As long as I'm responding, I am thinking this may need to be an optional build option. Perhaps with another `finaleplugin` flag, but I would prefer another mechanism if it is possible. At some point I may want to build into RGP Lua an info window similar to the one JW Lua has for each script, and I wouldn't want to have to guess what the format of the text is. I can't assume every user has `pandoc` either.

**Comment by rpatters1** [comment]
P.S. It might be possible to suppress the tabbing behavior by settin the edit focus to the Ok button, if you add one.

**Comment by asherber** [comment]
What if I put the RTF notes into a different property, like `finaleplugin.RtfNotes`? Then your info window could always rely on `Notes` being text (or markdown) and just ignore `RtfNotes`.

Setting focus to the Ok button works like a charm, thanks!

**Comment by rpatters1** [comment]
Call it `RTFNotes`, in keeping with `HashURL` where the acronym is all-caps.

**Comment by rpatters1** [comment]
I'm not sure what till happen with your test script on mac. (I've forgotten.) If you want me to try one, send it to me.

**Comment by asherber** [comment]
With the merging of #665 and #666, we now have:

- Whitespace in `finaleplugin.Notes` is not mangled
- Any text or Markdown in `Notes` is transformed to RTF in `finaleplugin.RTFNotes` in the `dist` script file
- Utility function `utils.show_notes_dialog()` will display a basic dialog with the contents of `RTFNotes`

@cv-on-hub, any need to keep this issue open?

**Comment by cv-on-hub** [comment]
Haven't had time to test it all yet but if all that is true then the issue is solidly closed.

**Comment by cv-on-hub** [comment]
RTFNotes looks great, and I realise this isn't rendering Markdown, so we can't retain nicely formatted bullet lists or tab spacing using `md` code. But that isn't the end of the world. I also discovered a problem [here](https://github.com/finale-lua/lua-scripts/blob/master/dist/rest_slider.lua#L5035) where some of the original list lacked 2 trailing spaces at the end of lines so the newline wasn't forced. (Hard to track that in the input phase).

Also - Arial 10-point is very squinty on my system. Is 12-point a reasonable compromise? And is it a problem hard-wiring coded RTF into  the `dist` files, making it hard to change later? Is it acceptable for authors to include our **own** RTFNotes for script rendering, alongside plugin.Notes for markdown display on the website?

**Comment by asherber** [comment]
The RTF is generated by pandoc directly from Notes -- if Notes uses Markdown, then pandoc will convert that into RTF. I've found that it handles bullets and numbered lists quite well, in addition to bold, italic, and headings, and inline and fenced code. I'll take a look at the example you pointed out to see why it didn't convert as expected.

As for font size, I don't think there will be something that looks ideal on all systems, given different OSes and monitor resolutions. On Windows, Arial 10 displays larger than the default text in a dialog box, so it seems fine to me, but I'm not opposed to changing it to some other reasonable default.

**Comment by asherber** [comment]
And if the source script already has an `RTFNotes`, then the build process will leave it alone.

**Comment by asherber** [comment]
Ah, it looks like the issue was not the trailing spaces on the list lines, but the fact that there was no blank line between "KEY COMMANDS" and the list.

**Comment by asherber** [comment]
@cv-on-hub You might want to consider using inline code or bold to indicate your keypresses, instead of brackets. I notice that you use `[x]`, but in Markdown that represents a completed task, and pandoc interprets it that way:

![image](https://github.com/finale-lua/lua-scripts/assets/5248041/dad42b61-bc5b-40dc-be81-21635d1416de)

**Comment by cv-on-hub** [comment]
Sorry if I'm slow on the uptake sometimes but I've sorted this out in my head and can now see how powerful this addition is to the repo - on top of curing the missing whitespace problem. 

In case it helps others, I found out it's really easy to install `pandoc` on MacOS, and not hard to get KeyboardMaestro to pass Markdown text to it (via the system clipboard) and get an economical RTF render ready to paste into `finaleplugin.RTFNotes`. Powerful!

**Comment by asherber** [comment]
One of the goals here was to make the process as transparent as possible, so that the auto-generated RTF gives a reasonable (and largely consistent) presentation of the content in almost all cases. If you've got suggestions for changing the defaults, then let's discuss.

RTF is an opaque and finicky format; if someone is going to be writing and tweaking Markdown in order to generate your own RTF, why not just use that Markdown in the `Notes` field and let the build process take care of the rest? Of course, if you want to generate RTF so that you can test the visual display locally, that's fine too, but in that case it's probably not necessary to commit the RTF with your script. (Among other things, if someone chooses to commit their own RTF, then they're responsible for keeping it in sync when fixing typos or making changes.)

**Comment by cv-on-hub** [comment]
The default font size in `utils.show_notes_dialog()` remains illegible for me so I've submitted a PR suggesting a fourth optional input parameter, fontsize.

So my initial drive was to force a larger font size in self-minted RTF. I like using literal tab chars for formatting (which seem to register in Fin-dialogs on both Win and Mac) and haven't found any way to represent those in Markdown, and `pandoc` replaces them with single spaces when converting to RTF. They still read when inserted into raw RTF and as long as this survives `/dist` packaging it is a solution. Other solutions happily considered.

I'm having some other odd renderings from Markdown in plugindef.Notes but will let you know if I can identify clear problem spots. It is handy to now have at least some mechanism to preview Markdown before packaging.

**Comment by cv-on-hub** [comment]
There final skerricks of this issue are now addressed in PR #676 and this can go to bed.
